### PR TITLE
Add new scripts to move review cursor by page

### LINKED ISSUE #13157: windows terminal: NVDA reads invisible lines when using pagers such as less and more
### Steps to reproduce:
1. launch windows terminal with wsl2 (tested with ubuntu 20.04)
2. run command which pages its output, for example `man test`
3. press nvda+numpad7, observe that NVDA navigated to the beginning of man output with command synopsis: press numpad9 a few times to hear "TEST(1)                                             User Commands                                             TEST(1)"
4. press space bar to scroll the page down, you may repeat a few times to see the full effect
5. press shift+numpad7 again to go to the top
6. try to read following lines, press numpad9

### Actual behavior:
NVDA reads the command output from the very beginning, as if no pagination appeared: "TEST(1)                                             User Commands                                             TEST(1)"
### Expected behavior:
NVDA should read only the last page displayed. It worked like that before and it also works like that when using wsl in windows console (without the new windows terminal)

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2021.3
#### Windows version:
windows 10 21H1 (os build 19043.1348)
#### Name and version of other software in use when reproducing the issue:
Microsoft windows terminal  1.11.2921.0
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
I think it worked in 2020.4
#### If NVDA add-ons are disabled, is your problem still occurring?
N/A
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
N/A

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Closes #13157 (given microsoft/terminal#13756). Alternative solution to #13671.

### Summary of the issue:
There is no way to move the review cursor by page. This is especially useful in terminals as a "page" can be defined as a screenful of text, solving #13157 without violating any other assumptions in NVDA and still permitting easy jump to the absolute top/bottom of the terminal buffer using the currently existing scripts.

### Description of how this pull request fixes the issue:
Added new scripts (bound to <kbd>NVDA</kbd>+<kbd>pageUp</kbd>/<kbd>NVDA</kbd>+<kbd>pageDown</kbd> on desktop and <kbd>NVDA</kbd>+<kbd>Shift</kbd>+<kbd>pageUp</kbd>/<kbd>NVDA</kbd>+<kbd>Shift</kbd>+<kbd>pageDown</kbd> on laptop) to move to previous/next page respectively.

### Testing strategy:
* Tested in UIA word where the page unit is sanely defined. Scripts work as expected.
* Tested in non-UIA word where the page unit is undefined. NVDA reports that movement by page is not supported.
* Tested in Notepad where the page unit is undefined but defers to document. NVDA reports "top"/"bottom" and reads the entire document as expected.

### Known issues with pull request:
As different `textInfo`s define `UNIT_PAGE` differently, behaviour will vary across applications.

### Change log entries:
Feel free to clarify/reword where you see fit.

=== New Features ===
- Introduced new commands to move the review cursor by page where supported by the application. These are especially useful for moving across screens of text, such as paginated output, in Windows Terminal.
    -  Move to previous page: <kbd>NVDA</kbd>+<kbd>pageUp</kbd> (desktop), <kbd>NVDA</kbd>+<kbd>Shift</kbd>+<kbd>pageUp</kbd> (laptop)
    -  Move to next page: <kbd>NVDA</kbd>+<kbd>pageDown</kbd> (desktop), <kbd>NVDA</kbd>+<kbd>Shift</kbd>+<kbd>pageDown</kbd> (laptop)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by codeofdusk** [comment]
CC @seanbudd, @leonardder, @michaelDCurran.

File: user_docs/en/userGuide.t2t [hunk 1]
```diff
@@ -630,6 +630,8 @@ The following commands are available for reviewing text:
 | Report current character in review | numpad2 | NVDA+. | none | Announces the current character on the line of text where the review cursor is positioned. Pressing twice reports a description or example of that character. Pressing three times reports the numeric value of the character in decimal and hexadecimal. |
 | Move to next character in review | numpad3 | NVDA+rightArrow | flick right (text mode) | Move the review cursor to the next character on the current line of text |
 | Move to end of line in review | shift+numpad3 | NVDA+end | none | Moves the review cursor to the end of the current line of text |
+| Move to previous page in review | NVDA+pageUp | NVDA+shift+pageUp | none | Moves the review cursor to the previous page of text if supported by the application |
+| Move to next page in review | NVDA+pageDown | NVDA+shift+pageDown | none | Moves the review cursor to the next page of text if supported by the application |
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:634]: Add code formatting to the shortcuts

```suggestion
| Move to previous page in review | ``NVDA+pageUp`` | ``NVDA+shift+pageUp`` | none | Moves the review cursor to the previous page of text if supported by the application |
| Move to next page in review | ``NVDA+pageDown`` | ``NVDA+shift+pageDown`` | none | Moves the review cursor to the next page of text if supported by the application |
```

File: source/globalCommands.py [hunk 2]
```diff
@@ -1433,6 +1433,61 @@ def script_review_nextLine(self, gesture):
...
 		speech.speakTextInfo(newLine, unit=textInfos.UNIT_LINE, reason=controlTypes.OutputReason.CARET)
 
+	@script(
+		# Translators: Input help mode message for move review cursor to previous page command.
+		description=_("Moves the review cursor to the previous page of the current navigator object and speaks it"),
+		resumeSayAllMode=sayAll.CURSOR.REVIEW,
+		category=SCRCAT_TEXTREVIEW,
+		gestures=("kb:NVDA+pageUp", "kb(laptop):NVDA+shift+pageUp", "ts(text):flickUp")
+	)
+	def script_review_previousPage(self, gesture):
```
Comment (seanbudd) [inline source/globalCommands.py:1443]:
```suggestion
	def script_review_previousPage(self, gesture: inputCore.InputGesture) -> None:
```

File: source/globalCommands.py [hunk 3]
```diff
@@ -1433,6 +1433,61 @@ def script_review_nextLine(self, gesture):
...
+		speech.speakTextInfo(info, unit=textInfos.UNIT_PAGE, reason=controlTypes.OutputReason.CARET)
+
+	@script(
+		# Translators: Input help mode message for move review cursor to next page command.
+		description=_("Moves the review cursor to the next page of the current navigator object and speaks it"),
+		resumeSayAllMode=sayAll.CURSOR.REVIEW,
+		category=SCRCAT_TEXTREVIEW,
+		gestures=("kb:NVDA+pageDown", "kb(laptop):NVDA+shift+pageDown", "ts(text):flickUp")
+	)
+	def script_review_nextPage(self, gesture):
```
Comment (seanbudd) [inline source/globalCommands.py:1468]:
```suggestion
	def script_review_nextPage(self, gesture: inputCore.InputGesture) -> None:
```

**Review by seanbudd — COMMENTED** [review]
I think this approach is a safe and generally useful feature.
The way it handles the different cases seems appropriate.

**Review by seanbudd — APPROVED** [review]
Great work @codeofdusk!

**Review by Qchristensen — APPROVED** [review]
User Guide reads well, great work.

File: user_docs/en/changes.t2t [hunk 4]
```diff
@@ -8,8 +8,18 @@ What's New in NVDA
 == New Features ==
 - Added a "Quick Start Guide" section to the User Guide. (#13934)
 - Introduced a new command to check the keyboard shortcut of the current focus. (#13960)
-  - Desktop: ``shift+numpad2``
-  - Laptop: ``NVDA+ctrl+shift+.``
+  - Desktop: ``shift+numpad2``.
+  - Laptop: ``NVDA+ctrl+shift+.``.
+  -
+- Introduced new commands to move the review cursor by page where supported by the application.
```
Comment (codeofdusk) [inline user_docs/en/changes.t2t:14]: @seanbudd I think it's worth calling out the utility of these commands in Windows Terminal as it isn't immediately obvious. Unless you think not for some reason?

Comment (seanbudd) [reply user_docs/en/changes.t2t:14]: Isn't a terminal change still blocking this?

I think this kind of extra information is better suited to the release blurb, in-process or in the WinConsole section of the user guide.

Comment (codeofdusk) [reply user_docs/en/changes.t2t:14]: > Isn't a terminal change still blocking this?

Yes, but it'll almost certainly be in by 2022.4.

Comment (codeofdusk) [reply user_docs/en/changes.t2t:14]: > WinConsole section of the user guide

Not applicable here, as these are for Terminal (but maybe a far future conhost).

Comment (seanbudd) [reply user_docs/en/changes.t2t:14]: I would suggest adding a note to the user guide, and add it to the changelog as well. This should probably wait until the terminal feature is introduced.

Comment (seanbudd) [reply user_docs/en/changes.t2t:14]: Perhaps a new section in the user guide then?

**Comment by burmancomp** [comment]
I tried these commands in Thunderbird 102.1.2 message list. NVDA + Shift + PageUp gave result "movement by page not supported" which is expected behavior I suppose. But command NVDA + Shift + PageDown resulted the following in log:

IO - inputCore.InputManager.executeGesture (10:28:16.276) - winInputHook (7160):
Input: kb(laptop):shift+NVDA+pageDown
ERROR - scriptHandler.executeScript (10:28:16.329) - MainThread (14408):
error executing script: <bound method GlobalCommands.script_review_nextPage of <globalCommands.GlobalCommands object at 0x0BF290F0>> with gesture 'vaihto+NVDA+sivu alas'
Traceback (most recent call last):
  File "scriptHandler.py", line 216, in executeScript
    script(gesture)
  File "globalCommands.py", line 1479, in script_review_nextPage
    newPage.expand(textInfos.UNIT_PAGE)
  File "textInfos\offsets.py", line 527, in expand
    self._startOffset,self._endOffset=self._getUnitOffsets(unit,self._startOffset)
  File "textInfos\offsets.py", line 505, in _getUnitOffsets
    raise ValueError("unknown unit: %s"%unit)
ValueError: unknown unit: page

**Comment by leonardder** [comment]
I can not reproduce this in the thunderbird messages list. I'm currently investigating what's necessary for broader rollout of page movement.

**Comment by codeofdusk** [comment]
@burmancomp I also cannot reproduce this, but have opened PR #14029 to remediate.

**Comment by codeofdusk** [comment]
@burmancomp Now that #14029 has been merged, could you please test the latest master snapshot and verify that NVDA now reports "movement by page not supported" for both previous and next?

**Pull request merged**